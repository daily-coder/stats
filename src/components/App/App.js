import { HashRouter, Route, Routes } from "react-router-dom";
import SharedLayout from "../../pages/SharedLayout";
import Home from "../../pages/Home";
import Overview from "../../pages/Overview";
import Color from "../../pages/Color";
import BgColor from "../../pages/BgColor";
import NotFound from "../../pages/NotFound";
import DataProvider from "../../context/DataContext";
import ThemeProvider from "../../context/ThemeContext";
import Category from "../../pages/Category";

function App() {
  return (
    <div>
      <DataProvider>
        <ThemeProvider>
          {/* github pages doesn't work with BrowserRouter */}
          <HashRouter>
            <Routes>
              <Route path="/" element={<Home />} />
              <Route path="/category" element={<SharedLayout />}>
                {/**
                 * overview, color and bg-color pages internal logic is different from other
                 * categories that's why we don't render them using useParams hook
                 */}
                <Route path="overview" element={<Overview />} />
                <Route path="color" element={<Color />} />
                <Route path="bg-color" element={<BgColor />} />
                <Route path=":id" element={<Category />} />
              </Route>
              <Route path="*" element={<NotFound />} />
            </Routes>
          </HashRouter>
        </ThemeProvider>
      </DataProvider>
    </div>
  );
}

export default App;
