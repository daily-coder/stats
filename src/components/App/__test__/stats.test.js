import {
  render as rtlRender,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { HashRouter } from "react-router-dom";

import App from "../App";

function render(ui, { route = "/", ...options } = {}) {
  window.history.pushState({}, "Test page", route);

  // BrowserRouter is not supported in github pages
  function Wrapper({ children }) {
    return <HashRouter>{children}</HashRouter>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...options });
}

test("navigate to overview page on clicking get started button", async () => {
  render(<App />);

  await userEvent.click(screen.getByRole("link", { name: /get started/i }));
  await waitForElementToBeRemoved(screen.getByLabelText(/loading/i));

  expect(
    screen.getByRole("heading", { name: /file size/i })
  ).toBeInTheDocument();
});

test("render new stats data on clicking search button", async () => {
  render(<App />, { route: "#/category/overview" });

  const searchBar = screen.getByRole("searchbox");

  await userEvent.clear(searchBar);
  await userEvent.type(searchBar, "github.com");
  await userEvent.click(screen.getByRole("button", { name: /search button/i }));
  await waitForElementToBeRemoved(screen.getByLabelText(/loading/i));

  expect(
    screen.getByRole("heading", { name: /file size/i })
  ).toBeInTheDocument();
});

test("display error message if url is empty", async () => {
  render(<App />, { route: "#/category/overview" });

  await userEvent.clear(screen.getByRole("searchbox"));
  await userEvent.click(screen.getByRole("button", { name: /search button/i }));

  // loading animation is not triggered if search value is empty and search button
  // is clicked. Instead error message is immediately displayed.

  // error message is big as it also provides some suggestions for user
  // check snapshot to know more about message.

  expect(screen.getByRole("alert")).toMatchSnapshot();
});

test("display error message if url is invalid", async () => {
  render(<App />, { route: "#/category/overview" });

  const searchBar = screen.getByRole("searchbox");

  await userEvent.clear(searchBar);
  await userEvent.type(searchBar, "abcdefgh");
  await userEvent.click(screen.getByRole("button", { name: /search button/i }));
  await waitForElementToBeRemoved(screen.getByLabelText(/loading/i));

  expect(screen.getByRole("alert")).toMatchSnapshot();
});
