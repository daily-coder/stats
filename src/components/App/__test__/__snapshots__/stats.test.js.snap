// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display error message if url is empty 1`] = `
<div
  role="alert"
>
  <h2>
    There was an Error.
  </h2>
  <ol
    class="list-inside list-decimal"
  >
    <li>
      Check your internet connection.
    </li>
    <li>
      Make sure url address is correct.
    </li>
  </ol>
</div>
`;

exports[`display error message if url is invalid 1`] = `
<div
  role="alert"
>
  <h2>
    There was an Error.
  </h2>
  <ol
    class="list-inside list-decimal"
  >
    <li>
      Check your internet connection.
    </li>
    <li>
      Make sure url address is correct.
    </li>
  </ol>
</div>
`;
